## Factbook Data Analysis SQL

  We will use SQL to analyze the factbook data and to understand the 'facts' table

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

Displaying all the tables that are available, we will be focusing on the 'facts' table

In [2]:
%%sql
SELECT *
FROM sqlite_master
WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

List the first 5 rows from the facts table

In [3]:
%%sql
SELECT *
FROM facts
LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

### List the Name, Minimum population, Maximum population, Minimum population growth, and Maximum population growth from the facts table


In [4]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

List the name of the Country that has the minimum population from the facts table


In [5]:
%%sql
SELECT name
FROM facts
WHERE population IN (SELECT MIN(population)
                     FROM facts)

 * sqlite:///factbook.db
Done.


[('Antarctica',)]

List the name of the Country that has the maximum population from the facts table


In [6]:
%%sql
SELECT name
FROM facts
WHERE population IN (SELECT MAX(population)
                     FROM facts)

 * sqlite:///factbook.db
Done.


[('World',)]

#### Antartica has a population of 0, while the country column had accidentally included the World as a country which has a population of 7,256,490,011

### Find the Max Pop. , Min Pop. , Min Pop. Growth, and Max Pop. Growth without the 'World'

*We will negate the 'World' row since it should not be included in the country column nor should it affect the data based upon the 'World' being considered a country*

In [7]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
FROM facts
WHERE name != 'World'

 * sqlite:///factbook.db
Done.


[(0, 1367485388, 0.0, 4.02)]

We will display the Average population and average area to understand the population density of each country

In [8]:
%%sql
SELECT AVG(population), AVG(area)
FROM facts
WHERE name != 'World'

 * sqlite:///factbook.db
Done.


[(32242666.56846473, 555093.546184739)]

### List the name of the countries that have a higher population than the average population and a lower area than the average area from the facts table


In [9]:
%%sql
SELECT name
FROM facts
WHERE population > (SELECT AVG(population)
                    FROM facts
                    WHERE name != 'World') AND
      area < (SELECT AVG(area)
                    FROM facts
                    WHERE name!= 'World')

 * sqlite:///factbook.db
Done.


[('Bangladesh',),
 ('Germany',),
 ('Iraq',),
 ('Italy',),
 ('Japan',),
 ('Korea, South',),
 ('Morocco',),
 ('Philippines',),
 ('Poland',),
 ('Spain',),
 ('Thailand',),
 ('Uganda',),
 ('United Kingdom',),
 ('Vietnam',)]

## Conclusion  
  These countries tend to be the most population dense countries due to the fact that they have a high population while having a lower area compared to the averages. 